src/vulkan-renderer/io/byte_stream.cpp

```diff
@@ -0,0 +1,112 @@
+#include "inexor/vulkan-renderer/io/byte_stream.hpp"
+#include "inexor/vulkan-renderer/world/cube.hpp"
+
+#include <fstream>
+
+namespace inexor::vulkan_renderer::io {
+std::vector<std::uint8_t> ByteStream::read_file(const std::filesystem::path &path) {
+    std::ifstream stream(path, std::ios::in | std::ios::binary);
+    return {std::istreambuf_iterator<char>(stream), std::istreambuf_iterator<char>()};
+}
+
+ByteStream::ByteStream(std::vector<std::uint8_t> buffer) : m_buffer(std::move(buffer)) {}
+
+ByteStream::ByteStream(const std::filesystem::path &path) : ByteStream(read_file(path)) {}
+
+std::size_t ByteStream::size() const {
+    return m_buffer.size();
+}
+[[nodiscard]] const std::vector<std::uint8_t> &ByteStream::buffer() const {
+    return m_buffer;
+}
+
+void ByteStreamReader::check_end(const std::size_t size) const {
+    if (std::distance(m_iter, m_stream.buffer().end()) < size) {
+        throw std::runtime_error("end would be overrun");
+    }
+}
+
+ByteStreamReader::ByteStreamReader(const ByteStream &stream) : m_stream(stream), m_iter(stream.buffer().begin()) {}
+
+void ByteStreamReader::skip(const std::size_t size) {
+    std::size_t skip = std::min(
+        size, std::size_t(std::distance<std::vector<std::uint8_t>::const_iterator>(m_iter, m_stream.buffer().end())));
+    std::advance(m_iter, skip);
+}
+
+std::size_t ByteStreamReader::remaining() const {
+    return std::distance<std::vector<std::uint8_t>::const_iterator>(m_iter, m_stream.buffer().end());
+}
+
+template <>
+std::uint8_t ByteStreamReader::read() {
+    check_end(1);
+    return *m_iter++;
+}
+
+template <>
+std::uint32_t ByteStreamReader::read() {
+    check_end(4);
+    return *m_iter++ | (*(m_iter++) << 8) | (*(m_iter++) << 16) | (*(m_iter++) << 24);
+}
+
+template <>
+std::string ByteStreamReader::read(const std::size_t &size) {
+    check_end(size);
+    auto start = m_iter;
+    std::advance(m_iter, size);
+    return std::string(start, m_iter);
+}
+
+template <>
+world::Cube::Type ByteStreamReader::read() {
+    return world::Cube::Type(read<std::uint8_t>());
```

Comment:
```suggestion
    return static_cast<world::Cube::Type>(read<std::uint8_t>());
```
Again, `static_cast` vs functional style cast.